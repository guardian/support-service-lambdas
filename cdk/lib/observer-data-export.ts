import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack, GuStringParameter } from '@guardian/cdk/lib/constructs/core';
import {
	type GuFunctionProps,
	GuLambdaFunction,
} from '@guardian/cdk/lib/constructs/lambda';
import { type App, Duration } from 'aws-cdk-lib';
import { ArnPrincipal, Role, User } from 'aws-cdk-lib/aws-iam';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { nodeVersion } from './node-version';

export class ObserverDataExport extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const app = 'observer-data-export';

		const bucket = new Bucket(this, 'Bucket', {
			bucketName: `${app}-${this.stage.toLowerCase()}`,
			lifecycleRules: [{ expiration: Duration.days(28) }],
		});

		const unifidaUser = new User(this, 'UnifidaUser', {
			userName: `unifida-${this.stage.toLowerCase()}`,
		});

		bucket.grantRead(unifidaUser);

		const airflowCloudComposerUserArnParameter = new GuStringParameter(
			this,
			`${app}-airflow-cloud-composer-user-arn`,
			{
				description: `Airflow cloud composer user ARN (Ophan Account)`,
			},
		);

		bucket.grantReadWrite(
			new ArnPrincipal(airflowCloudComposerUserArnParameter.valueAsString),
			`Observer_newsletter_eligible/*`,
		);

		const lambdaDefaultConfig: Pick<
			GuFunctionProps,
			'app' | 'memorySize' | 'fileName' | 'runtime' | 'timeout' | 'environment'
		> = {
			app,
			memorySize: 1024,
			fileName: `${app}.zip`,
			runtime: nodeVersion,
			timeout: Duration.seconds(300),
			environment: { Stage: this.stage },
		};

		new GuLambdaFunction(this, 'ObserverDataExportPlaceholderLambda', {
			...lambdaDefaultConfig,
			handler: 'observerDataExportPlaceholder.handler',
			functionName: `observer-data-export-placeholder-${this.stage}`,
		});
	}
}
