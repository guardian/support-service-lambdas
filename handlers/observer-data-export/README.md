# Observer Data Export

## Overview

The **ObserverDataExport** project is designed to automate the process of exporting Observer-only subscriber data from Salesforce and securely encrypting it for further processing. The data is transferred from Salesforce to an S3 bucket and encrypted using AES-256-CBC encryption before being uploaded to a shared S3 bucket. The encryption ensures that only authorized users, such as **Unifida** (Tortoise's dev team), can access and decrypt the data. This project is scheduled to run every Tuesday at 7 AM UTC.

## Architecture

The architecture involves multiple AWS resources, including:

**AWS S3 Buckets**: For data storage, including the storage of encrypted files and manifests.
**AWS AppFlow**: For extracting data from the Observer Subscriber Data custom object in Salesforce into a CSV file
- **AWS Lambda**: For encrypting the Salesforce data and uploading it to the shared bucket.
- **AWS SQS**: For handling events triggered when new data is uploaded to the Salesforce data transfer bucket.
- **CloudWatch Alarms**: For monitoring and alerting in case of failures.

![architecture diagram](./docs/architecture-diagram.svg)
[Link to Google Draw Diagram](https://docs.google.com/drawings/d/10A6zS81k6fmk4meViSiY732NbZEKu5wM0R5J2--H5Ys/edit)

### Data Flow

1. **Salesforce Data Transfer**: Salesforce Observer subscriber data is exported weekly (every Tuesday at 7 AM UTC) using AWS AppFlow and transferred to an S3 bucket, **SalesforceObserverDataTransferBucket**.
2. **File Upload and Encryption**:
   - Once new data is uploaded to **SalesforceObserverDataTransferBucket**, an SQS event triggers an AWS Lambda function.
   - The Lambda function downloads the unencrypted data (CSV file) and encrypts it using AES-256-CBC.
   - The encryption key (AES) is itself encrypted using RSA, a public key from Unifida.
3. **Upload to Shared Bucket**:

   - The encrypted AES key, IV, and CSV data are uploaded to a **shared S3 bucket**, **UnifidaSharedBucketName**.
   - A manifest file is created for each upload that includes the paths to the encrypted AES key, IV, and CSV file, along with the MD5 hash of the CSV file for verification.

4. **Permission Management**:
   - **Unifida's IAM User**: An IAM user for Unifida (Tortoise's dev team) is created to access the files in the shared bucket. The user is able to decrypt the AES key using their RSA private key and use it to decrypt the data.
   - **DataTech IAM User**: Permissions are granted to the **DataTech IAM user** (Airflow Cloud Composer user from the AWS Ophan account) to write data to the **Observer_newsletter_eligible/** folder, which contains subscriber data for other products in addition to the Observer-only data.
5. **Bucket Expiration**:
   - **Shared Bucket**: The shared bucket has a **28-day expiration policy** for the data stored. It also has full CloudTrail enabled, including S3 data events, to monitor access to the bucket.
   - **Intermediate Bucket**: The intermediate bucket (used for unencrypted data) has a **1-day retention period** due to the sensitive nature of the data.

## Security and Access

1. **Unifida's Decryption Process**:
   - Unifida will be able to **decrypt the AES key** using their RSA private key.
   - They will then use the decrypted AES key and IV to **decrypt the CSV file** containing the Observer subscriber data.
   - To ensure the data integrity and that encryption was successful, Unifida can **verify the MD5 hash** of the CSV file against the provided MD5 signature in the manifest.

## Lambda Function

The Lambda function used for encryption and uploading data is triggered by an SQS event and performs the following actions:

- Fetches the public RSA key from the shared S3 bucket.
- Encrypts the AES key using RSA encryption.
- Encrypts the CSV file using AES-256-CBC encryption.
- Uploads the encrypted AES key, IV, and CSV file to the shared bucket.
- Creates a manifest file with the paths to the encrypted files and the MD5 hash of the CSV file.

## CloudWatch Alarm

A **CloudWatch Alarm** is set up to monitor the Lambda function's execution. If there are issues with the Lambda function (e.g., failure to process data), the alarm will trigger and alert the responsible team. The alarm monitors the **dead-letter queue** for failed Lambda invocations.

## Weekly Schedule

The Salesforce data extraction process is scheduled to run every **Tuesday at 7 AM UTC**, as defined in the cron expression `cron(0 7 ? * TUE *)`.

## Permissions Overview

- **Unifida IAM User**:
  - Can decrypt the AES key using their RSA key.
  - Can decrypt the CSV file using AES-256-CBC and the AES IV.
  - Can verify the encryption by comparing the MD5 hash in the manifest with the actual file.
- **DataTech IAM User**:
  - Has permissions to write to the **Observer_newsletter_eligible/** folder for the Observer data in subscriptions that include other products, not just Observer-only data.

## Conclusion

The **ObserverDataExport** project ensures secure handling and transfer of Observer subscriber data, providing Unifida with the tools to decrypt and verify the data while maintaining full audit and security controls on the AWS infrastructure.
