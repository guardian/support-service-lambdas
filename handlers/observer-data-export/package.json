{
	"name": "observer-data-export",
	"description": "Application in charge of exporting Observer data.",
	"scripts": {
		"test": "jest --group=-integration",
		"it-test": "jest --group=integration",
		"build": "esbuild --bundle --platform=node --target=node20 --outdir=target src/handlers/*.ts",
		"lint": "eslint src/**/*.ts test/**/*.ts",
		"package": "pnpm type-check && pnpm check-formatting && pnpm test && pnpm build && cd target && zip -qr observer-data-export.zip ./*.js",
		"type-check": "tsc --noEmit",
		"check-formatting": "prettier --check **.ts",
		"fix-formatting": "prettier --write **.ts"
	},
	"dependencies": {}
}
