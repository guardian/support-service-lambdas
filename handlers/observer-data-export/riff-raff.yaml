stacks:
  - support
regions:
  - eu-west-1
allowedStages:
  - CODE
  - PROD
deployments:
  observer-data-export-cloudformation:
    type: cloud-formation
    app: observer-data-export
    parameters:
      templateStagePaths:
        CODE: observer-data-export-CODE.template.json
        PROD: observer-data-export-PROD.template.json
  observer-data-export:
    type: aws-lambda
    parameters:
      fileName: observer-data-export.zip
      bucketSsmLookup: true
      prefixStack: false
      functionNames:
        - placeholder-
    dependencies: [observer-data-export-cloudformation]
