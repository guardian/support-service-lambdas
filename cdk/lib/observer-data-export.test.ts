import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { ObserverDataExport } from './observer-data-export';

describe('The ObserverDataExport stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const codeStack = new ObserverDataExport(app, `observer-data-export-CODE`, {
			stack: 'support',
			stage: 'CODE',
		});
		const prodStack = new ObserverDataExport(app, `observer-data-export-PROD`, {
			stack: 'support',
			stage: 'PROD',
		});
		expect(Template.fromStack(codeStack).toJSON()).toMatchSnapshot();
		expect(Template.fromStack(prodStack).toJSON()).toMatchSnapshot();
	});
});
