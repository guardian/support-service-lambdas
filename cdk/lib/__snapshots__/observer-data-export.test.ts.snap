// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ObserverDataExport stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuStringParameter",
      "GuDistributionBucketParameter",
      "GuLambdaFunction",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "observerdataexportairflowcloudcomposeruserarn": {
      "Description": "Airflow cloud composer user ARN (Ophan Account)",
      "Type": "String",
    },
  },
  "Resources": {
    "Bucket83908E77": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": "observer-data-export-code",
        "LifecycleConfiguration": {
          "Rules": [
            {
              "ExpirationInDays": 28,
              "Status": "Enabled",
            },
          ],
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "BucketPolicyE9A3008A": {
      "Properties": {
        "Bucket": {
          "Ref": "Bucket83908E77",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Ref": "observerdataexportairflowcloudcomposeruserarn",
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "Bucket83908E77",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "Bucket83908E77",
                          "Arn",
                        ],
                      },
                      "/Observer_newsletter_eligible/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Ref": "observerdataexportairflowcloudcomposeruserarn",
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "Bucket83908E77",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "Bucket83908E77",
                          "Arn",
                        ],
                      },
                      "/Public_keys/unifida_public_rsa_key.pem",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "ObserverDataExportPlaceholderLambdaE320A2A9": {
      "DependsOn": [
        "ObserverDataExportPlaceholderLambdaServiceRoleDefaultPolicy801551DE",
        "ObserverDataExportPlaceholderLambdaServiceRole134BD8E9",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "support/CODE/observer-data-export/observer-data-export.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "observer-data-export",
            "STACK": "support",
            "STAGE": "CODE",
            "Stage": "CODE",
          },
        },
        "FunctionName": "observer-data-export-placeholder-CODE",
        "Handler": "observerDataExportPlaceholder.handler",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "ObserverDataExportPlaceholderLambdaServiceRole134BD8E9",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "observer-data-export",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ObserverDataExportPlaceholderLambdaServiceRole134BD8E9": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "observer-data-export",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ObserverDataExportPlaceholderLambdaServiceRoleDefaultPolicy801551DE": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/support/CODE/observer-data-export/observer-data-export.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/support/observer-data-export",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/support/observer-data-export/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ObserverDataExportPlaceholderLambdaServiceRoleDefaultPolicy801551DE",
        "Roles": [
          {
            "Ref": "ObserverDataExportPlaceholderLambdaServiceRole134BD8E9",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "UnifidaUserA0A8377C": {
      "Properties": {
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
        "UserName": "unifida-code",
      },
      "Type": "AWS::IAM::User",
    },
    "UnifidaUserDefaultPolicy85929A11": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "Bucket83908E77",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "Bucket83908E77",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "UnifidaUserDefaultPolicy85929A11",
        "Users": [
          {
            "Ref": "UnifidaUserA0A8377C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;

exports[`The ObserverDataExport stack matches the snapshot 2`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuStringParameter",
      "GuDistributionBucketParameter",
      "GuLambdaFunction",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "observerdataexportairflowcloudcomposeruserarn": {
      "Description": "Airflow cloud composer user ARN (Ophan Account)",
      "Type": "String",
    },
  },
  "Resources": {
    "Bucket83908E77": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": "observer-data-export-prod",
        "LifecycleConfiguration": {
          "Rules": [
            {
              "ExpirationInDays": 28,
              "Status": "Enabled",
            },
          ],
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "BucketPolicyE9A3008A": {
      "Properties": {
        "Bucket": {
          "Ref": "Bucket83908E77",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Ref": "observerdataexportairflowcloudcomposeruserarn",
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "Bucket83908E77",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "Bucket83908E77",
                          "Arn",
                        ],
                      },
                      "/Observer_newsletter_eligible/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Ref": "observerdataexportairflowcloudcomposeruserarn",
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "Bucket83908E77",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "Bucket83908E77",
                          "Arn",
                        ],
                      },
                      "/Public_keys/unifida_public_rsa_key.pem",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "ObserverDataExportPlaceholderLambdaE320A2A9": {
      "DependsOn": [
        "ObserverDataExportPlaceholderLambdaServiceRoleDefaultPolicy801551DE",
        "ObserverDataExportPlaceholderLambdaServiceRole134BD8E9",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "support/PROD/observer-data-export/observer-data-export.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "observer-data-export",
            "STACK": "support",
            "STAGE": "PROD",
            "Stage": "PROD",
          },
        },
        "FunctionName": "observer-data-export-placeholder-PROD",
        "Handler": "observerDataExportPlaceholder.handler",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "ObserverDataExportPlaceholderLambdaServiceRole134BD8E9",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "observer-data-export",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ObserverDataExportPlaceholderLambdaServiceRole134BD8E9": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "observer-data-export",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ObserverDataExportPlaceholderLambdaServiceRoleDefaultPolicy801551DE": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/support/PROD/observer-data-export/observer-data-export.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/support/observer-data-export",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/support/observer-data-export/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ObserverDataExportPlaceholderLambdaServiceRoleDefaultPolicy801551DE",
        "Roles": [
          {
            "Ref": "ObserverDataExportPlaceholderLambdaServiceRole134BD8E9",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "UnifidaUserA0A8377C": {
      "Properties": {
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "UserName": "unifida-prod",
      },
      "Type": "AWS::IAM::User",
    },
    "UnifidaUserDefaultPolicy85929A11": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "Bucket83908E77",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "Bucket83908E77",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "UnifidaUserDefaultPolicy85929A11",
        "Users": [
          {
            "Ref": "UnifidaUserA0A8377C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
